features/mbedtls/targets/TARGET_STM/aes_alt.c

```diff
@@ -0,0 +1,914 @@
+                       const unsigned char *key,
+                       unsigned int keybits)
+{
+    unsigned int i;
+    int ret = 0;
+
+#if (MBEDTLS_DEBUG)
+    printf("[ALT] aes_set_key\n");
+#endif
+
+    AES_VALIDATE_RET(ctx != NULL);
+    AES_VALIDATE_RET(key != NULL);
+
+    /* Protect context access                                  */
+    /* (it may occur at a same time in a threaded environment) */
+#if defined(MBEDTLS_THREADING_C)
+    if (mbedtls_mutex_lock(&cryp_mutex) != 0) {
+        return (MBEDTLS_ERR_THREADING_MUTEX_ERROR);
+    }
+#endif /* MBEDTLS_THREADING_C */
+
+    /* include the appropriate instance name */
+#if defined (AES)
+    ctx->hcryp_aes.Instance = AES;
+#elif defined (AES1)
+    ctx->hcryp_aes.Instance = AES1;
+#else /* CRYP */
+    ctx->hcryp_aes.Instance = CRYP;
+#endif /* AES */
+
+    switch (keybits) {
+        case 128:
+            ctx->hcryp_aes.Init.KeySize = CRYP_KEYSIZE_128B;
+            break;
+
+        case 192:
+#if ( USE_AES_KEY192 == 1 )
+            ctx->hcryp_aes.Init.KeySize = CRYP_KEYSIZE_192B;
+            break;
+#else
+            ret = MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED;
+            goto exit;
+#endif /* USE_AES_KEY192 */
+
+        case 256:
+            ctx->hcryp_aes.Init.KeySize = CRYP_KEYSIZE_256B;
+            break;
+
+        default :
+            ret = MBEDTLS_ERR_AES_INVALID_KEY_LENGTH;
+            goto exit;
+    }
+
+    /* Format and fill AES key  */
+    for (i = 0; i < (keybits / 32); i++) {
+        GET_UINT32_BE(ctx->aes_key[i], key, 4 * i);
+    }
+
+    ctx->hcryp_aes.Init.DataType = CRYP_DATATYPE_8B;
+    ctx->hcryp_aes.Init.DataWidthUnit = CRYP_DATAWIDTHUNIT_BYTE;
+    ctx->hcryp_aes.Init.pKey = ctx->aes_key;
+
+    if (HAL_CRYP_Init(&ctx->hcryp_aes) != HAL_OK) {
+        ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+        goto exit;
+    }
+
+    /* allow multi-context of CRYP : save context */
+    ctx->ctx_save_cr = ctx->hcryp_aes.Instance->CR;
+
+exit :
+    /* Free context access */
+#if defined(MBEDTLS_THREADING_C)
+    if (mbedtls_mutex_unlock(&cryp_mutex) != 0) {
+        ret = MBEDTLS_ERR_THREADING_MUTEX_ERROR;
+    }
+#endif /* MBEDTLS_THREADING_C */
+
+    return (ret);
+}
+
+void mbedtls_aes_init(mbedtls_aes_context *ctx)
+{
+    AES_VALIDATE(ctx != NULL);
+
+    __disable_irq();
+#if defined(MBEDTLS_THREADING_C)
+    /* mutex cannot be initialized twice */
+    if (!cryp_mutex_started) {
+        mbedtls_mutex_init(&cryp_mutex);
+        cryp_mutex_started = 1;
+    }
+#endif /* MBEDTLS_THREADING_C */
+    cryp_context_count++;
+    __enable_irq();
+
+    cryp_zeroize((void *)ctx, sizeof(mbedtls_aes_context));
+
+#if (MBEDTLS_DEBUG)
+    printf("[ALT] mbedtls_aes_init %u\n", cryp_context_count);
+#endif
+}
+
+
+void mbedtls_aes_free(mbedtls_aes_context *ctx)
+{
+#if (MBEDTLS_DEBUG)
+    printf("[ALT] mbedtls_aes_free %u\n", cryp_context_count);
+#endif
+
+    if (ctx == NULL) {
+        return;
+    }
+
+    __disable_irq();
```

Comment:
This is a question of re-entrance.
Depending of the SW architecture, this code can be simultaneously used from different places (several threads, single thread, even within an interrupt).
Nevertheless, one cryptoprocessor HW instance is available and must be shared.
Even in a single thread, user can call mbedtls primitives in different sequences and all this processing can be interrupted.
For all these reasons, disabling IRQ seems to be the most appropriate solution to deal with all these situations.

Reply:
If we are for sure using this from an interrupt, sure, I agree we need a stronger primitive. However, AES in an interrupt? What's the use case for that? I suspect just thread safety would be enough for crypto operations, which generally take a long time. As crypto operations do take a long time, you certainly don't want to either do them either within an interrupt context or to block all other interrupts during the crypto operation.

Reply:
I wouldn't expect any mbed TLS crypto calls to be IRQ safe - not sure it's been defined anywhere.

With the current code, the "mutex init" inside the disable_irq would not be legal, if that was leading to an actual RTOS osMutexInit - that can't be called either from IRQ or with IRQs disabled, so the current form doesn't make sense.

I'd also be very much inclined to avoid the complexity by not attempting the juggling - just write in C++ and use a `SingletonPtr<Mutex>`. That would have the mutex be statically allocated (if referenced) and initialised on first use.

Reply:
What can we use to ensure thread safety, knowing that we're trying to release the mutex here that is used at other places to ensure this thread safety ? 
Also we're looking for a C implementation that can be used as well outside of mbed-os ...

Reply:
-Of course, crypto code would not be called within an interrupt (it cannot because it is done with HAL cryp used in polling mode and this implementation will takes too long time in this case. There is no Use Case yet). Sorry for the noise.
-Anyway, you said “mutex_init” (or “mutex_free)” inside the “disable_irq/enable_irq” would not be legal. “mutex_init”/ “mutex_free” are functions with basic processing (chained lists, …). They would not cause a context switch. I do not see what is wrong in calling them with IRQs disabled. Could you explain please?
-“cryp_context_count” and “cryp_mutex_started” should remain valid variables (“atomicity”). I mean more than one instruction can be completed and a switch to another task should not occur until the instructions are executed. What is the right method? Is there a mbed-OS specific function to protect them? 
-As pointed by @LMESTM, what would be the C implementation because this code is also used in mbed TLS environment.

Reply:
`osMutexNew` from IRQ isn't legal because the CMSIS RTOS docs say so.

In practice, it's because RTX uses supervisor calls both for memory protection and serialisation, and for any call to be usable from interrupt or with interrupts disabled on Cortex-M they need an alternative non-supervisor-call form to be manually coded. That's only done for the calls that are needed from IRQ, like `osSemaphoreRelease`

It's a cute "feature" of the ARM-M architecture that the `SVC` instruction causes a Hard Fault if issued with interrupts disabled. Quite different to ARM-A.

An increment of `cryp_context_count` won't be atomic in hardware, because it's "load", "add", "store" - three instructions. Breaks if there's a task switch between the load and store, and someone else increments or decrements in the meantime. A plain load or store can be, but you still have potential compiler reordering issues, plus the fact you're performing a complete operation - the mutex init - it's not just one access.

If you're trying to write embedded-portable C code here, I don't think there's a portable construct for atomic operations. C11 and C++11 both have them, but implementations aren't always ideal in embedded. Mbed OS has `mstd::atomic`, which is a local implemenation of `std::atomic`, but other platforms won't have that. So I think atomics are out.

In Mbed OS I'd use `singleton_lock()` to protect your reference count and mutex creation. Is there any equivalent in mbed TLS porting layers? Or is there any place you could guarantee a central init call to init your master mutex?

`std::mutex` guarantees auto-init on first `lock` - without that you basically need a single init entry point, else you start needing a mutex to protect your mutex creation, and you're stuck in a loop.